File: 1-chapter/Part_3/toweOfHanoiSolver.js
let TOTAL_DISKS = 6;
let TOWERS = {
    'A': [],
    'B': [],
    'C': []
}

for (let i = TOTAL_DISKS; i > 0; i--) {
    TOWERS['A'].push(i);
}

function printDisk(diskNum) {
    let emptySpace = ' '.repeat(TOTAL_DISKS - diskNum);
    if (diskNum === 0){
        console.log(`${emptySpace} || ${emptySpace}`);
    }else {
        let diskSpace = "@".repeat(diskNum);
        let diskNumLabel = String(' ' + diskNum).slice(-2);
        console.log(emptySpace + diskSpace + diskNumLabel + diskSpace + emptySpace)
    }
}

function printTowers() {
    let towerLetter = 'ABC';
    for (let level = TOTAL_DISKS; level >=0; level--){
        for (let towerLetterIndex = 0; towerLetterIndex < 3; towerLetterIndex++){
            let tower = TOWERS[towerLetter[towerLetterIndex]]
            if (level >= tower.length){
                printDisk(0);
            }else {
                printDisk(tower[level])
            }
        }
        console.log();
    }
    let emptySpace = ' '.repeat(TOTAL_DISKS);
    console.log(`${emptySpace} A ${emptySpace} ${emptySpace} B ${emptySpace} ${emptySpace}`)
    console.log('='.repeat(TOTAL_DISKS * 6)); // Разделитель
}

function moveOneDisk(startTower, endTower) {
    let disk = TOWERS[startTower].pop();
    TOWERS[endTower].push(disk);
}

function solve(numberOfDisks, startTower, endTower, tempTower) {
    if (numberOfDisks === 1){
        moveOneDisk(startTower, endTower);
        printTowers()
        return;
    }else {
        solve(numberOfDisks - 1, startTower, tempTower, endTower)
        moveOneDisk(startTower, endTower);
        printTowers();
        solve(numberOfDisks - 1, tempTower, endTower, startTower);
        return;
    }
}

printTowers();
solve(TOTAL_DISKS, 'A', 'B', 'C')

File: 1-chapter/Part_3/toweOfHanoiSolver.py
import sys

TOTAL_DISKS = 6

TOWERS = {
    'A': list(reversed(range(1, TOTAL_DISKS + 1))),
    'B': [],
    'C': []
}


def printDisk(diskNum):
    emptySpace = ' ' * (TOTAL_DISKS - diskNum)
    if diskNum == 0:
        sys.stdout.write(emptySpace + '||' + emptySpace)
    else:
        diskSpace = '@' * diskNum
        diskNumLabel = str(diskNum).rjust(2, '_')
        sys.stdout.write(emptySpace + diskSpace + diskNumLabel + diskSpace + emptySpace)


def printTowers():
    for level in range(TOTAL_DISKS, -1, -1):
        for tower in (TOWERS['A'], TOWERS['B'], TOWERS['C']):
            if level >= len(tower):
                printDisk(0)
            else:
                printDisk(tower[level])
        sys.stdout.write('\n')
    emptySpace = ' ' * (TOTAL_DISKS)
    print('%s A%s%s B%s%s C\n' % (emptySpace, emptySpace, emptySpace, emptySpace, emptySpace))


def moveOneDisk(startTower, endTower):
    disk = TOWERS[startTower].pop()
    TOWERS[endTower].append(disk)


def solve(numberOfDisks, startTower, endTower, tempTower):
    if numberOfDisks == 1:
        moveOneDisk(startTower, endTower)
        printTowers()
        return
    else:
        solve(numberOfDisks - 1, startTower, tempTower, endTower)
        moveOneDisk(startTower, endTower)
        printTowers()
        solve(numberOfDisks - 1, tempTower, endTower, startTower)
        return


printTowers()
solve(TOTAL_DISKS, 'A', 'B', 'C')

# while True:
#     printTowers()
#     print('Enter letter of start tower the end tower. (A,B,C) Or Q to quit.')
#     move = input().upper()
#     if move == 'Q':
#         sys.exit()
#     elif move[0] in 'ABC' and move[1] in 'ABC' and move[0] != move[1]:
#         moveOneDisk(move[0], move[1])
